### DISCUSSION TITLE
How to update session on server side?

### DISCUSSION THREAD

**Discussion by kAutoman** [nextauthjs/next-auth#7864] [Help]
Currently I'm trying to update sever side session.
I've used external api when login so jwt comming from external API and getting user information from that token on nextjs server side.
The problem is after change some info in external api side, I need refresh session info on nextjs server side.
But when calling this function,
```
const { update } = useSession();
update();

```

that's not updating nextjs jwt token.
I need update jwt token.

Here's the code part to set user info from external api.
```
app\api\auth\[...nextauth]\route.ts

callbacks: {
  async session({ session, token, user }) {
      // Send properties to the client, like an access_token and user id from a provider.
      session.accessToken = token.accessToken;
      session.user.userRole = token.userRole;
      session.user.info = token.userInfo;
      return session;
  },
}

providers: [
    CredentialsProvider({
      id: "credentials",
      name: "email",
      credentials: {
        username: {
          label: "Email Or Phone number",
          type: "text",
          placeholder: "example@example.com",
        },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        try {
          const user = await login({
            email: credentials?.username,
            password: credentials?.password,
          });

          const res = await getDataBasic("/users/me?token=" + user.token);

          res.token = user.token;
          res.tokenExpires = user.tokenExpires
          return res;
        } catch (ex: AxiosError) {
          const error =
            ex.response.data.message || "Wrong username or password";
          return Promise.reject(new Error(error));
        }
      },
    }),
  ],

```

```
 const user = await login({
            email: credentials?.username,
            password: credentials?.password,
 });
```
This is login part using external API.
Best regards.

**Comment by ryancwalsh** [comment]
I asked the same question here: https://stackoverflow.com/q/77396220/470749

**Comment by Wundero** [comment]
I think you can use [the JWT callback](https://authjs.dev/guides/basics/callbacks#jwt-callback) in your auth settings in order to mutate the `token` (which stores your user data) if you need to modify the JWT itself, but otherwise you can just add what you need to the `session` callback. A note on the difference: `session` just sends a JSON object to the client, but doesn't save the data you add to it anywhere (since you aren't using a DB), while `jwt` actually persists the data on the client side with the cookie storing the JWT.

**Reply by ryancwalsh** [reply]
Thanks, but the API endpoint of mine that receives the 2nd access code is _outside_ of next-auth, since trying to let a user login with a 2nd provider would overwrite the 1st.

So I don't know how to pass my value to the JWT callback from there.

**Reply by Wundero** [reply]
I'm not sure if there's an ergonomic way to do that. I suppose you could try to store the access token in-memory somewhere before accessing the user's session (e.g. with `getServerSession`), in a way that would be accessible to the callback, and then you could use that stored value in the callback. This might be a bit awkward, and actually possibly risky (if the same server instance is replying to multiple different users) since you could accidentally introduce token sharing.

I don't think there is an easy way to update JWT token data on the server side, though I haven't used JWTs much so I can't be 100% certain. From my brief searching of the codebase, there doesn't seem to be anything you can do to accomplish this.

Might be worth making a PR or issue about this, since updating sessions is a pretty nice thing to be able to do server-side, and is trivial with databases.

**Comment by itsbetma** [comment]
NextAuth is just another API route. What you can do is use the useSession.update with a parameter to identify an update process.

```javascript
// React
const { update } = useSession();
update({forceUpdate:true});
```

Then on [...nextauth]:

Write an API logic before calling the NextAuth handler.

The logic that I wrote will update your useSession.data and the cookie that stores the next session, doing so your changes will be held during page refreshes. 

**FYI: JWT_SECRET is the one used on the authOptions.**

```javascript
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
  // Checking if session.update was called with data={forceUpdate:true}
  if (req?.body?.data?.forceUpdate === true) {
    const cookieName = !req?.headers?.referer?.includes('https://')
      ? 'next-auth.session-token'
      : '__Secure-next-auth.session-token';
    const session = await getServerSession(req, res, authOptions);
    // Checking if session is valid
    if (session !== null && session.user?.email) {
      // Retrieving the next-auth.session-token from the cookies
      let token = req.headers.cookie
        ?.split(';')
        .map((t) => t.trim())
        .filter((t) => t.startsWith(`${cookieName}=`))[0];
      if (token) {
        token = token.replace(`${cookieName}=`, '');
        // Token is encoded. We need to use the decode from next auth
        // to retrieve the token as an object
        const tokenParsed = await decode({
          token,
          secret: String(process.env.JWT_SECRET)
        });
        // Checking tokenParsed, it can be null
        if (tokenParsed) {
          const newData = 'anything'
          tokenParsed['newData'] = newData;
          const tokenEncoded = await encode({
            token: tokenParsed,
            secret: String(process.env.JWT_SECRET)
          });
          setCookie({ res }, cookieName, tokenEncoded, {
            path: '/',
            maxAge: 60 * 60 * 24 * 30, // 30 days
            httpOnly: true,
            sameSite: 'Lax',
            secure: true
          });

          return res.status(200).json({ ...session, newData });
        }
      }
    }
  }

  return await NextAuth(req, res, authOptions);
}
```

**Reply by ryancwalsh** [reply]
Interesting. It's true that I haven't tried decrypting, editing, and then re-encrypting the token directly. I could try that. Thanks.
